### PR TITLE
Prefix message length for control messages.

### LINKED ISSUE #281: Support length prefixed control message

### PR DESCRIPTION
**Pull request by GhostofCookie**
Closes #281

### CODE REVIEW

File: src/messages.cpp [hunk 1]
```diff
@@ -141,59 +137,30 @@ namespace quicr::messages {
         buffer << UintVar(msg.last_group_id);
         buffer << UintVar(msg.last_object_id);
 
+        UintVar size(buffer.size());
+        buffer.insert(buffer.begin(), size.begin(), size.end());
+
```
Comment (suhasHere) [inline src/messages.cpp:142]: wouldn't this causes reshuffling of the elements ?

File: src/messages.cpp [hunk 2]
```diff
@@ -972,77 +778,41 @@ namespace quicr::messages {
         PushExtensions(buffer, msg.extensions);
         buffer << UintVar(msg.payload.size());
         buffer << msg.payload;
+
+        UintVar size(buffer.size());
+        buffer.insert(buffer.begin(), size.begin(), size.end());
```
Comment (suhasHere) [inline src/messages.cpp:783]: Non control message doesn't need length update

File: src/messages.cpp [hunk 3]
```diff
@@ -972,77 +778,41 @@ namespace quicr::messages {
...
         return buffer;
     }
 
     template<class StreamBufferType>
     bool operator>>(StreamBufferType& buffer, MoqObjectDatagram& msg)
     {
-        switch (msg.current_pos) {
```
Comment (suhasHere) [inline src/messages.cpp:981]: ObjectDatagram is not a control message

File: src/messages.cpp [hunk 4]
```diff
@@ -1056,40 +826,22 @@ namespace quicr::messages {
         buffer << UintVar(msg.subscribe_id);
         buffer << UintVar(msg.track_alias);
         buffer << UintVar(msg.priority);
+
+        UintVar size(buffer.size());
+        buffer.insert(buffer.begin(), size.begin(), size.end());
```
Comment (suhasHere) [inline src/messages.cpp:831]: Not a control message

File: src/messages.cpp [hunk 5]
```diff
@@ -1056,40 +826,22 @@ namespace quicr::messages {
...
         return buffer;
     }
 
     template<class StreamBufferType>
     bool operator>>(StreamBufferType& buffer, MoqStreamHeaderTrack& msg)
```
Comment (suhasHere) [inline src/messages.cpp:836]: Not a control message:

File: test/moq_messages.cpp [hunk 6]
```diff
@@ -58,6 +58,23 @@ Verify(std::vector<uint8_t>& buffer, uint64_t message_type, T& message, [[maybe_
...
+    StreamBuffer<uint8_t> in_buffer;
+    in_buffer.Push(buffer);
+
+    auto length = in_buffer.DecodeUintV();
+    CHECK(length.has_value());
+    CHECK_EQ(*length, in_buffer.Size());
+
+    auto msg_type = in_buffer.DecodeUintV();
+    CHECK(msg_type.has_value());
+    CHECK_EQ(*msg_type, message_type);
```
Comment (suhasHere) [inline test/moq_messages.cpp:74]: I think the order is wrong here .
Its Type, Length and body in that order.

**Review by suhasHere — CHANGES_REQUESTED** [review]
Thanks for the PR;

- This PR updates non control messages. Non control messages are not updated for length prefix
- Tests seems to look at wrong order

On something nice to do , but can wait is  to avoid reshuflfing memory by doing
- assume length can be encoded in 2 bytes and process the buffer. 
- if length exceed 2 bytes, then reallocate the buffer.
Most of the control message lengths defined today can be defined in 2 bytes.

File: src/messages.cpp (hunk 1 above)
Comment (GhostofCookie) [reply src/messages.cpp:142]: Yep! Thank UintVar for that.

Comment (TimEvens) [reply src/messages.cpp:142]: Would have been nice to be fixed value instead of uintvar for header length.  @suhasHere , do you have another way to do it?

**Review by TimEvens — COMMENTED** [review]
Had meeting with Tomas. Changes are coming that should address the questions as well as improve processing using header length.

**Review by TimEvens — APPROVED** [review]
I really like the changes.

**Review by suhasHere — APPROVED** [review]

**Pull request merged**